testplan/runners/pools/connection.py

```diff
@@ -2,46 +2,348 @@
+            if received is None:
+                raise RuntimeError('Received None when {} was expected.'.format(
+                    expect))
+            assert received.cmd == expect
+        return received
+
+
+class QueueClient(Client):
+    """Queue based client implementation, for thread pool workers to communicate with its pool."""
+
+    def __init__(self, recv_sleep=0.05):
+        super(QueueClient, self).__init__()
+        self._recv_sleep = recv_sleep
+        self.requests = None
+        self.responses = []     # single-producer(pool) single-consumer(worker) FIFO queue
+
+    def connect(self, requests):
+        """
+        Connect to the request queue of Pool
+        :param requests: request queue of pool that worker should write to.
+        :type requests: Queue
+        """
+        self.requests = requests
+        self.active = True
+
+    def disconnect(self):
+        """Disconnect worker from pool"""
+        self.active = False
+        self.requests = None
+
+    def send(self, message):
+        """
+        Worker sends a message
+        :param message: Message to be sent.
+        :type message: :py:class:`~testplan.runners.pools.communication.Message`
+        """
+        if self.active:
+            self.requests.put(message)
+
+    def receive(self):
+        """
+        Worker receives response to the message sent, this method blocks.
+        :return: Response to the message sent.
+        :type: :py:class:`~testplan.runners.pools.communication.Message`
+        """
+        while self.active:
+            try:
+                return self.responses.pop()
+            except IndexError:
+                time.sleep(self._recv_sleep)
+
+    def respond(self, message):
+        """
+        Used by :py:class:`~testplan.runners.pools.base.Pool` to respond to
+        worker request.
+
+        :param message: Respond message.
+        :type message: :py:class:`~testplan.runners.pools.communication.Message`
+        """
+        if self.active:
+            self.responses.append(message)
+
+
+class ZMQClient(Client):
+    """
+    ZMQ based client implementation for process worker to communicate with its pool.
+    :param address: Pool server address to connect to.
+    :type address: ``float``
+    :param recv_sleep: Sleep duration in msg receive loop.
+    :type recv_sleep: ``float``
+    """
+
+    def __init__(self, address, recv_sleep=0.05, recv_timeout=5):
+        super(ZMQClient, self).__init__()
+        self._address = address
+        self._recv_sleep = recv_sleep
+        self._recv_timeout = recv_timeout
+        self._context = None
+        self._sock = None
+
+        self.connect()  # auto connect
+
+    def connect(self):
+        """Connect to a ZMQ Server"""
+        self._context = zmq.Context()
+        self._sock = self._context.socket(zmq.REQ)
+        self._sock.connect("tcp://{}".format(self._address))
+        self.active = True
+
+    def disconnect(self):
```

Comment:
Could we add a `__del__` method to ensure that the sockets are always closed as a fallback even if `disconnect()` has not been called - like we have for ZMQServer?